tests/test_vertical_portscans.py

```diff
@@ -208,3 +227,307 @@ def test_check_if_enough_dports_to_trigger_an_evidence(
+    timewindow = 'timewindow0'
+    dstip = '8.8.8.8'
+    amount_of_dports = 10
+    cache_key = vertical_ps.get_cache_key(profileid, timewindow, dstip)
+    vertical_ps.cached_tw_thresholds[cache_key] = amount_of_dports
+
+    enough = vertical_ps.check_if_enough_dports_to_trigger_an_evidence(cache_key, amount_of_dports)
+
+    assert not enough
+
+
+def test_get_cache_key_empty_values(mock_db):
+    vertical_ps = ModuleFactory().create_vertical_portscan_obj(mock_db)
+    profileid = ""
+    twid = ""
+    dstip = ""
+
+    cache_key = vertical_ps.get_cache_key(profileid, twid, dstip)
+
+    assert cache_key == "::dstip::VerticalPortscan"
+    
+def test_combine_evidence_multiple_keys(mock_db):
+    vertical_ps = ModuleFactory().create_vertical_portscan_obj(mock_db)
+    vertical_ps.pending_vertical_ps_evidence = {
+        "profile_1.1.1.1-timewindow0-Not Established-TCP-8.8.8.8": [
+            (1700828217.314165, 100, ["uid1", "uid2"], 10),
+            (1700828217.314165, 200, ["uid3", "uid4"], 10),
+            (1700828217.314165, 300, ["uid5", "uid6"], 10)
+        ],
+        "profile_2.2.2.2-timewindow1-Not Established-UDP-9.9.9.9": [
+            (1700828217.314165, 400, ["uid7", "uid8"], 20),
+            (1700828217.314165, 500, ["uid9", "uid10"], 20),
+            (1700828217.314165, 600, ["uid11", "uid12"], 20)
+        ]
+    }
+
+    mock_set_evidence = mock_db.set_evidence
+    vertical_ps.combine_evidence()
+    assert mock_set_evidence.call_count == 2
+
+
+def test_decide_if_time_to_set_evidence_or_combine_new_cache_key(mock_db):
+    vertical_ps = ModuleFactory().create_vertical_portscan_obj(mock_db)
+    evidence = {
+        'timestamp': 1700828217.314165,
+        'pkts_sent': 100,
+        'protocol': 'TCP',
+        'profileid': 'profile_1.1.1.1',
+        'twid': 'timewindow0',
+        'uid': ['uid1', 'uid2'],
+        'amount_of_dports': 10,
+        'dstip': '8.8.8.8',
+        'state': 'Not Established',
+    }
+    cache_key = 'new_cache_key'
+
+    mock_set_evidence = mock_db.set_evidence
+
+    enough = vertical_ps.decide_if_time_to_set_evidence_or_combine(evidence, cache_key)
+
+    assert enough
+    assert vertical_ps.alerted_once_vertical_ps[cache_key]
+    mock_set_evidence.assert_called_once()
+    
+def test_check_no_connections(mock_db):
+    vertical_ps = ModuleFactory().create_vertical_portscan_obj(mock_db)
+    profileid = "profile_1.1.1.1"
+    twid = "timewindow0"
+
+    mock_db.get_data_from_profile_tw.return_value = {}
+    mock_set_evidence = mock_db.set_evidence
+
+    vertical_ps.check(profileid, twid)
+
+    mock_set_evidence.assert_not_called()
+    
+def test_get_not_established_dst_ips_invalid_protocol(mock_db):
+    vertical_ps = ModuleFactory().create_vertical_portscan_obj(mock_db)
+    profileid = "profile_1.1.1.1"
+    twid = "timewindow0"
+    protocol = "INVALID"
+    state = "Not Established"
+    role = "Client"
+    type_data = "IPs"
+    direction = "Dst"
+
```

Comment:
The above 3 lines are unused local variables. usually your IDe should warn you about them or mark them in gray.

Reply:
Yes, it is actually getting displayed in a light color. Sorry, I missed that earlier. I am fixing it.